### ISSUE TITLE
git lfs migrate import on path doesn't handle files that have been moved

### ISSUE THREAD

**Issue by amaidment** [git-lfs/git-lfs#4167]
During the course of using `git lfs migrate import [path]/**` (on a `--mirror clone` of our repo), I found it gave me some odd behaviour where files had been moved between directories in the repository - for example, if the file was in com/xyz/abc/big_file.csv, but at some point in time was moved to com/xyz/def/big_file.csv.

**Migrate Current Path**
If I do `git lfs migrate import com/xyz/def/** --verbose`, it does _not_ output that the file (com/xyz/def/big_file.csv) has been moved to LFS. While it adds com/xyz/def/** to .gitattributes, it doesn't actually move the file to LFS storage. As such, if I then do a (non-mirror) clone from my mirror repo, I get the following error:
> message:Encountered 1 file(s) that should have been pointers, but weren't:
>         com/xyz/def/big_file.csv

**Migrate Original Path**
Instead, if I do the `git lfs migrate import com/xyz/abc/** --verbose`, it outputs that the file (com/xyz/abc/big_file.csv) has been moved to LFS. 

> migrate: Sorting commits: ..., done.
> migrate: commit [blah]: com/xyz/abc/big_file.csv
> migrate: Rewriting commits: 100% (x/x), done.

So, it adds com/xyz/abc/** to .gitattributres, and moves the file to LFS storage. However, it doesn't seem to handle that the file has been moved correctly. As such, if I then do a (non-mirror) clone from my mirror repo, there are no error messages, but the repo's state seems inconsistent.That is that .gitattributes contains the original path:

> com/xyz/abc/** filter=lfs diff=lfs merge=lfs -text

However, if I look at what files are stored under LFS, it shows the file at its new path.

> $ git lfs ls-filesblah - com/xyz/def/big_file.csv

I'm not quite sure what the desired behaviour should be, in either of these cases, but what it's actually doing in both these cases seems confusingly inconsistent.

**Workarounds**
I found two workarounds for this problem:

1. Rather than doing the LFS migration for files in specific locations, I broadened the migration (e.g. across test resources) to a higher directory - i.e. equivalent to doing git lfs migrate import com/xyz/**

2. Where 1) was neither possible or appropriate (i.e. the path change was at high level directory, e.g. between modules), I found that I could do both the lfs migration on both directories (in the appropriate order) - i.e.

> $ git lfs migrate import com/xyz/abc/** 
> ...
> $ git lfs migrate import com/xyz/def/**  
> ...
> 

This seemed to get me to state I wanted to be in, with the files in LFS at the right time, and the right paths being tracked by LFS.

**Comment by chrisd8088** [comment]
Hey, thanks for writing this up; it does seem like inconsistent and/or confusing behaviour, where at a minimum it would be better if there could be some warning or error message instead.

However, I'm struggling a bit to replicate the behaviour you describe, mostly because `git lfs migrate import` takes branch refs, not paths, as arguments (see the [man page](https://github.com/git-lfs/git-lfs/blob/02f92e2f1c53eb771875ac01b49551e786604ded/docs/man/git-lfs-migrate.1.ronn#synopsis) for a fuller description).

Were you possibly using `--include=/com/xyz/...` arguments instead of bare paths?  Or perhaps you were using the `--no-rewrite` option, which [does take path arguments](https://github.com/git-lfs/git-lfs/blob/02f92e2f1c53eb771875ac01b49551e786604ded/docs/man/git-lfs-migrate.1.ronn#import-no-rewrite), but has different effects?

However, when I try to create a simple test repo with one commit adding a file in `foo/abc/file.bin` and a commit moving it to `foo/def/file.bin`, and then use `git lfs migrate import --include=foo/abc/**` or `--include=foo/def/**`, I get pretty much what I would expect, namely, a `.gitattributes` that lists the directory path I gave to `--include` and `file.bin` replaced by an LFS pointer file but *only* in the commits where it exists in the given file path.  So if I ran `git lfs migrate import --include=foo/abc/**` the file is an LFS pointer file in its initial creation commit, but is turned back into a full file in the commit which moved it to `foo/def/file.bin` because that's not in the pathspec now in `.gitattributes`.  And conversely if I ran `git lfs migrate import --include=foo/bar/**` then the file is only an LFS pointer file in the second commit where it's moved from its earlier location.

Would you be able to create a simple repo and set of steps which reproduce the behaviour you're seeing, and share those?  And it would also be useful to see the output `git lfs env` on your system, too.

**Comment by chrisd8088** [comment]
I'm going to close this for now, but if someone can reproduce it and provide a test case, by all means please re-open it!

**Comment by chrisd8088** [comment]
I'm reopening this after some investigation based on #4272, because I have found a fairly simple example test case which I think reproduces this problem.  (This may also be the source of some of the problems reported in #3238.)

Specifically, when using `git lfs migrate import --include=<path>`, if the `--include` path matches the location into which a directory (more precisely, a tree object) was moved, and specifically matches the name of the parent directory of the moved directory (as well as possibly the directory itself and/or its contents), then Git LFS appears to not actually include the blobs within the moved directory in its migration, although it does add the include path to the `.gitattributes` file as if it performed the migration.

Here's the test case, showing that only the file moved from `a1/` to `a2/` gets matched and migrated, while the one moved from `b1/` to `/b2/b1/` does not:
```
#!/bin/bash

git init rewrite
cd rewrite

mkdir -p a1 b1
echo foo > a1/foo.txt
echo bar > b1/bar.txt
git add .
git commit -m initial

mv a1 a2
git add .
git commit -m move1

mkdir b2
mv b1 b2
git add .
git commit -m move2

git lfs migrate import --include=/a2,/b2
echo
git lfs ls-files
```

which generates the output from `git lfs ls-files`:
```
b5bb9d8014 - a2/foo.txt
```

although it writes both `--include` paths into `.gitattributes`:
```
/a2 filter=lfs diff=lfs merge=lfs -text
/b2 filter=lfs diff=lfs merge=lfs -text
```

As noted, the `a2/` directory is matched and migrated, but there is no parent directory (other than the root tree) for it.  But the `b2/b1/` directory doesn't match against the `/b2` include path, although it should, apparently because the name of the parent directory into which `b1/` was moved (i.e., `/b2/`) is part of the include path.

More investigation will obviously be needed to clarify if this accurately captures the failure mode, and what the root cause of that failure mode is.

**Comment by chrisd8088** [comment]
This behaviour may have been resolved, at least in part, by #4671.  For example, the test case in the comment above now generates:
```
$ git lfs ls-files
b5bb9d8014 - a2/foo.txt
7d865e959b - b2/b1/bar.txt
```

**Comment by chrisd8088** [comment]
I believe this has been resolved more efficiently by #4813 (see https://github.com/git-lfs/git-lfs/pull/4813#pullrequestreview-853931058), so I'm going to close it again.
